File: ios/VergiPro/Core/Sync/SubmissionStatusPolling.swift
import Foundation

struct RemoteSubmissionStatus: Sendable {
    let state: QueuedMutation.State
    let progress: Double?
    let serverDocumentID: String?
    let retryAfter: TimeInterval?
}

protocol SubmissionStatusTransport: Sendable {
    func fetch(submissionID: UUID, organizationID: Int) async throws -> RemoteSubmissionStatus
}

actor SubmissionStatusPoller {
    private let queue: OfflineMutationQueue
    private let transport: any SubmissionStatusTransport

    init(queue: OfflineMutationQueue, transport: any SubmissionStatusTransport) {
        self.queue = queue; self.transport = transport
    }

    func poll(submissionID: UUID, organizationID: Int) async throws {
        var transportFailures = 0
        while !Task.isCancelled {
            do {
                let remote = try await transport.fetch(submissionID: submissionID, organizationID: organizationID)
                try await queue.transition(id: submissionID, to: remote.state, progress: remote.progress, serverDocumentID: remote.serverDocumentID)
                transportFailures = 0
                if remote.state.isTerminal { return }
                try await Task.sleep(for: .seconds(min(max(remote.retryAfter ?? 3, 1), 30)))
            } catch is CancellationError { throw CancellationError() }
            catch is SubmissionStateError { throw SubmissionStateError.invalidTransition }
            catch {
                transportFailures += 1
                let seconds = min(pow(2, Double(transportFailures)), 30) + Double.random(in: 0...1)
                try await Task.sleep(for: .seconds(seconds))
            }
        }
    }
}

extension QueuedMutation.State {
    var isTerminal: Bool { self == .needsVerification || self == .needsAttention || self == .completed }
}

File: android/app/src/main/java/com/vergipro/mobile/core/sync/SubmissionStatusPolling.kt
package com.vergipro.mobile.core.sync

import kotlinx.coroutines.delay
import java.util.UUID
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

data class RemoteSubmissionStatus(
    val state: MutationState,
    val progress: Float? = null,
    val serverDocumentId: String? = null,
    val retryAfterSeconds: Long? = null,
)

fun interface SubmissionStatusTransport {
    suspend fun fetch(submissionId: UUID, organizationId: Int): RemoteSubmissionStatus
}

class SubmissionStatusPoller(
    private val queue: OfflineMutationQueue,
    private val transport: SubmissionStatusTransport,
) {
    suspend fun poll(submissionId: UUID, organizationId: Int) {
        var failures = 0
        while (true) {
            try {
                val remote = transport.fetch(submissionId, organizationId)
                queue.transition(submissionId, remote.state, remote.progress, remote.serverDocumentId)
                failures = 0
                if (remote.state.isTerminal()) return
                delay((remote.retryAfterSeconds ?: 3).coerceIn(1, 30) * 1000)
            } catch (cancelled: kotlinx.coroutines.CancellationException) {
                throw cancelled
            } catch (state: SubmissionStateFailure) {
                throw state
            } catch (_: Exception) {
                failures++
                val seconds = min(2.0.pow(failures.toDouble()), 30.0) + Random.nextDouble(0.0, 1.0)
                delay((seconds * 1000).toLong())
            }
        }
    }
}

fun MutationState.isTerminal() = this == MutationState.NeedsVerification || this == MutationState.NeedsAttention || this == MutationState.Completed
